haskell.compiler.ghc*: cross-compiled ghc [iteration on alex's PR]

sternenseemann opened the pull request:
## Description of changes

Based on #243619

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AlexandreTunstall committed:
haskell.compiler.ghc*: set stage 0 tools

This is easy in comparison since these tools won't end up in GHC's
settings nor need to be available at runtime, so we can use
the *_FOR_BUILD environment variables.

It is important to add buildCC to depsBuildBuild to engage the
stdenv/wrapper script machinery properly.

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

AlexandreTunstall committed:
haskell.compiler.ghc*: fall back to host libs for “cross native” ghc

The goal of this commit is basically to eliminate the use of
targetPackages for finding libraries. Instead, we introduce a
`targetLibs` set that can be used instead. The libraries in there
philosophically come from targetPackages since they are used by the core
libs and will be linked against user code. However, when cross compiling
GHC it's always a native compiler, so we can and have to use
pkgsHostTarget (targetPackages would be empty). This is explained more
in the acccompanying comment.

An alternative to this approach is not to pass in the libraries
explicitly via `--with-*` flags and rely on cc-wrapper and splicing to
pick the correct library. This works well for ncurses and probably
merits testing for other libraries as well since it's very simple. It
would need to be verified, however, that configure doesn't discover the
“wrong” library and leaks it somewhere.

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

AlexandreTunstall committed:
haskell.compiler.ghc*: correctly account for ncurses in cross

1. Explicitly set WITH_TERMINFO. We usually match GHC's behavior well,
   but it is better to tie the Nix option to make explicitly.
   Unfortunately, the same is very complicated to achieve with
   hadrian (iirc).
2. Disable enableTerminfo if we are cross-compiling. This matches
   the behavior of GHC's build system, so we'll have to match it now.
   It also reduces the ncurses-related headache a bit.
3. Stop passing --with-curses* flags. Unfortunately, GHC does not
   account for the fact that different platforms need different ncurses
   libraries. This is somewhat migitated by the fact that ncurses is
   only ever needed for the build platform if we are cross compiling,
   but I seem to remember it leaking into the final GHC somehow.
   A more reliable alternative is relying on the cc/ld wrapper scripts,
   as they'll always pull out the correct ncurses out of the environment
   when GHC's build system passes -lcurses.
4. Unconditionally add ncurses to depsBuildBuild. Stage0 unconditionally
   builds terminfo (maybe the stage1 compiler needs it?), so we need to
   make sure that ncurses for the build platform is available.

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

AlexandreTunstall committed:
haskell.compiler.ghc*: don't configure --host

AlexandreTunstall committed:
haskell.compiler.ghc*: use host->target CC and tools in settings

GHC's build system assumes that the C compiler, tools etc. discovered
during configure can also be used at runtime. This means that the CC,
LD, AR etc. variables given at runtime are used to populate the settings
file which GHC uses to lookup the tools it needs.

The implicit assumption of this mechanism is that the build and runtime
environment of GHC are similar enough and PATH is used to find the
tools. I. e. if we set CC=clang, we wouldn't need to worry about this as
much. We, however, pass absolute paths which is useful since it allows
GHC to work outside of stdenv (as long as e. g. no FFI is involved).

Even so, until now, we didn't really have any problems stemming from
this, as we used pkgsBuildTarget to get everything we need. The
compiler we'd want to execute would in principle need to come
from pkgsHostTarget.

1. For native compilers, all package sets are the same since
   build == host == target.
2. For cross compilers build == host, so pkgsBuildTarget
   is practically the same as pkgsHostTarget.

When cross-compiling a native compiler, build != host, so we need to
actually ensure that GHC uses different tools at runtime compared to
bootstrapping. There is currently no intended way to achieve this, so we
use a custom tool to edit the settings file. An alternative would be to
patch the build system, but this would be difficult to maintain. We
could go down this route if there's interest from upstream to provide a
proper way to specify the runtime tools.

Co-authored-by: sternenseemann <sternenseemann@systemli.org>

AlexandreTunstall committed:
haskell.compiler.ghc*: don't cross-compile haddock

AlexandreTunstall committed:
haskell.compiler.ghc*: add enableUnregisterised

AlexandreTunstall committed:
haskell.compiler.ghc*: assert cross is possible

AlexandreTunstall committed:
haskell.compiler.ghc*: allow all platforms

Propagating the supported platforms of the boot compiler doesn't make
much sense when unregistered cross-compilation is possible.

AlexandreTunstall committed:
haskell.compiler.ghc*: use GHC from pkgsBuildBuild

AlexandreTunstall committed:
tests.cross.sanity: Add GHC to catch regressions

This is to ensure that Haskell users on platforms that lack official
bindists still have a convenient means of getting GHC running natively.

In my admittedly somewhat limited testing on RISC-V, GHC 8.10.7 is able
to bootstrap native builds for 9.2.8 and 9.4.5. GHC 9.2.8 and 9.4.5 are
unable to bootstrap themselves and 9.6.2 when cross-compiled.

If you're looking at this commit to see whether you can safely upgrade
the compiler used here to remove 8.10, please try cross-compiling 9.0 or
later and then booting a native GHC with it.

sternenseemann committed:
ghc-settings-edit: init at 0.1.0

AlexandreTunstall commented:
I like the change to use ghc-settings-edit, but would you mind adding a comment inside the LHS file and/or the meta description to briefly explain what the script is used for?

As mentioned in [this comment of the original PR](https://github.com/NixOS/nixpkgs/pull/243619#issuecomment-2037727394), the settings file apparently still references the the build platform's LLVM.
Have you had a chance to look into this issue?
I assume it just needs an extra rule to overwrite the LLVM field.

I'll test `pkgsCross.riscv64` and report back in a few days.

sternenseemann commented:
> but would you mind adding a comment inside the LHS file and/or the meta description to briefly explain what the script is used for?

Yes, I'll do that.

> still references the the build platform's LLVM

Forgot about that, I'll look into that. This is basically the case the old asserts guarded against?

```
assert targetCC == pkgsHostTarget.targetPackages.stdenv.cc;
assert buildTargetLlvmPackages.llvm == llvmPackages.llvm;
```

AlexandreTunstall commented:
> This is basically the case the old asserts guarded against?

I think so.

AlexandreTunstall commented:
[I have successfully used this PR's branch to boot a native compiler and build nix-output-monitor and ShellCheck on RISC-V.](https://github.com/AlexandreTunstall/nixos-riscv/tree/df3a550f99766ae7bfe88449494648cdb5f65e53)

The GHC 9.4.8 cross-build still needs to be unregisterised, but through the rebuild I was able to obtain a fully functional native RISC-V compiler, so there has been no regression on that front.

Since a cross-compiled 9.4.8 can be used to boot a native 9.6.5, please feel free to update the cross sanity test to build 9.4.8 instead of 8.10.7.
I can't remember why my past testing led me to believe that only 8.10.7 was usable as a boot compiler, but it no longer seems to be an issue.

AlexandreTunstall commented:
This error is probably what led me to believe that only 8.10 and 9.0 were usable as boot compilers:

> ghc.mk:141: *** The make build system requires a boot compiler older than ghc-9.2. Your boot compiler is too new and cannot be used to build ghc-9.4 with make. Either boot with ghc 9.0.2 or build with hadrian. See https://www.haskell.org/ghc/blog/20220805-make-to-hadrian.html for advice on transitioning to hadrian..  Stop.

At the time, Nixpkgs had not yet switched to 9.6.

sternenseemann commented:
Added some commentary to the ghc-settings-edit.lhs file.

sternenseemann commented:
- [ ] port everything to ghc-settings-edit
- [ ] cross splicing for libdw
- [ ] fix unlit in settings

maralorn commented on pkgs/development/tools/haskell/ghc-settings-edit/ghc-settings-edit.lhs:
```diff
@@ -0,0 +1,109 @@
+This tool was originally written by sternenseemann and is licensed under the MIT
+license (as is nixpkgs) as well as the BSD 3 Clause license since it incorporates
+some code from GHC. It is primarily intended for use in nixpkgs, so it should be
+considered unstabile: No guarantees about the stability of its command line
```
```suggestion
considered unstable: No guarantees about the stability of its command line
```

bolives-hax commented:
I was trying this out for s390x, while there were no evaluation errors or anything it won't build due to llvm related issues. But that hardly is the fault of the PR it looks pretty solid

sternenseemann committed:
haskell.compiler.ghc96: fix incorrect unlit path for cross compilers

Hadrian started installing unlit with a targetPrefix (if applicable)
which wasn't the case with make before. Unfortunately, the logic to
generate the settings file wasn't updated, so GHC 9.6.* cross compilers
expect to find an unlit binary without a target prefix.

Upstream issue: https://gitlab.haskell.org/ghc/ghc/-/issues/23317

sternenseemann committed:
haskell.compiler.ghc9{6,8,10}: update comment about stage selection

The build platform doesn't matter for checking which stage is our final
stage! Stage2 means host and target are sufficiently similar, Stage1
means host and target may differ.

sternenseemann committed:
haskell.compiler.*: set cctools tools in settings if applicable

sternenseemann committed:
haskell.compiler.*: set correct runtime LLVM tools in settings

Note that the clang command should be adapted in the future as done for
CLANG in #338782.

sternenseemann commented:
@AlexandreTunstall Does `enableDocs` imply `enableHaddock`? Is that why you made the condition for the former more strict?

sternenseemann commented:
```
/nix/store/6pmlfcq95bswq8vwb64wzasp9pk8pnm2-ghc-9.8.2
/nix/store/fqmmy8qahfh7yhpm9v4xfddfw56w8y2a-ghc-9.8.2-doc
/nix/store/zg2lzsx6ni1s6lm0wvznf2bl1lasii9v-ghc-aarch64-unknown-linux-gnu-8.10.7-doc
/nix/store/znjbb57b6jfq1kqcg67rcw4wvv3yszn9-ghc-aarch64-unknown-linux-gnu-8.10.7
/nix/store/7cmg5rfmnr07mms8fami2qkwbbfpswpv-aarch64-unknown-linux-gnu-ghc-9.6.6-doc
/nix/store/cam5dzipi1yq3y9nfx64mwr8vm6nhlkk-aarch64-unknown-linux-gnu-ghc-9.6.6
/nix/store/8pg837wpi3k6x82i715bkj6hg072ngv8-ghc-aarch64-unknown-linux-gnu-9.4.8-doc
/nix/store/wg0mppma2z6spmj6np32kzk2xh1qlrh9-ghc-aarch64-unknown-linux-gnu-9.4.8
```

Successfully compile.

sternenseemann commented:
- [x] Successfully (natively) compiled test programs with GHC 8.10.7, 9.4.8 cross compiled from x86_64-linux on aarch64-linux.
- [x] “Ordinary” Cross compilation from x86_64 to aarch64-linux with 9.6.6 still works.
- [x] Native 9.8.2 compiler still works.

AlexandreTunstall commented:
> Does enableDocs imply enableHaddock? Is that why you made the condition for the former more strict?

It's been a while since I wrote that change, so I can't remember.
One would have to try `enableDocs = true; enableHaddock = false` and see, because I can't find anything in the GHC documentation that suggests that one implies the other.

AlexandreTunstall commented:
On the latest version of this PR, `pkgsCross.riscv64.haskell.compiler.native-bignum.ghc948` builds, runs natively on my RV64GC_Zba_Zbb system, and outputs working RV64GC binaries.

sternenseemann commented on pkgs/test/cross/default.nix:
```diff
@@ -146,6 +146,8 @@ let
     pkgs.pkgsCross.mips64el-linux-gnuabin32.stdenv
     pkgs.pkgsCross.mingwW64.stdenv
+    # Unlike later versions, cross-compiled 8.10.7 can reliably bootstrap a native build
+    pkgs.pkgsCross.riscv64.haskell.compiler.integer-simple.ghc8107
```
Maybe an idea, I think we can expand the test list a bit (in release-haskell.nix), but should keep it to the essential stuff in the test.

The nice thing about 8.10.7 is of course that it can be used to start a bootstrap chain of all compilers in nixpkgs.

Also, we should be able to use gmp now if gmp can be cross compiled to riscv64.

sternenseemann commented:
> One would have to try `enableDocs = true; enableHaddock = false` and see, because I can't find anything in the GHC documentation that suggests that one implies the other.

I've confirmed that it only influences the manual build which isn't problematic when cross-compiling GHC. We only need/should disable it if it is a hassle to build sphinx for the host platform (e.g. as is the case with musl).
